Add blog post for Provision volumes from cross namespace data sources (#37628)
kubernetes/website: content/en/blog/_posts/2022-12-6-provision-volumes-from-cross-namespace-snapshots-alpha.md

---
layout: blog

> title: "Kubernetes 1.26: alpha support for provision volumes from cross-namespace snapshots"

Reviewer: ```suggestion
title: "Kubernetes v1.26: alpha support for provision volumes from cross-namespace snapshots"
```

date: 2022-12-6
slug: provision-volumes-from-cross-namespace-snapshots-redesigned
---

**Authors:**
Takafumi Takahashi (Hitachi Vantara)

Kubernetes v1.26, released earlier this month, introduced an enable usage of provision of persisten volume claim from volume snapshot in other namespaces.  

> Before Kubernetes v1.25, by using volume snapshots feature, users can provision volumes from snapshots. However, it only works for the `VolumeSnapshot` in the samme namespace, therefore users can't provision a persisten volume claim in one namespace from a `VolumeSnapshot` in other namespace. On the other hand, there are use cases that require to share the `VolumeSnapshot` across namespaces.  

Reviewer 2: Before Kubernetes v1.26

Author: Fixed.

> Before Kubernetes v1.25, by using volume snapshots feature, users can provision volumes from snapshots. However, it only works for the `VolumeSnapshot` in the samme namespace, therefore users can't provision a persisten volume claim in one namespace from a `VolumeSnapshot` in other namespace. On the other hand, there are use cases that require to share the `VolumeSnapshot` across namespaces.  

Reviewer: nit:

```suggestion
Before Kubernetes v1.25, by using volume snapshots feature, users can provision volumes from snapshots. However, it only works for the `VolumeSnapshot` in the same namespace, therefore users can't provision a persisten volume claim in one namespace from a `VolumeSnapshot` in other namespace. On the other hand, there are use cases that require to share the `VolumeSnapshot` across namespaces.  
```

>  To solve this problem,Kubernetes v1.26 includes a new API field called `dataSourceRef2`.

Reviewer 2: I think this blog needs to be updated to reflect the changes in the implementation, i.e., it is dataSourceRef that is updated to include a namespace and there isn't a dataSourceRef2.

Author: Fixed.

>  To solve this problem,Kubernetes v1.26 includes a new API field called `dataSourceRef2`.

Author: Fixed.

## How it works

> Once the csi-provisioner finds a VolumeSnapshot is specified with non-empty namespace as dataSourceRef2, it checks all ReferenceGrants in　PersistentVolumeClaim.spec.dataSourceRef2.namespace to see if access to the snapshot is allowed. If it is allowed, the csi-provisioner provisions a volume from the snapshot.  

Reviewer 3: We should recap what CSI is, with a hyperlink to read more.

I recommend mentioning / explaining what a ReferenceGrant is at this point, and then:

```suggestion
Once the csi-provisioner finds that a VolumeSnapshot is specified with a `dataSourceRef02` that
has a non-empty namespace name,
it checks all reference grants within the namespace that's specified by the `.spec.dataSourceRef2.namespace`
field of the PersistentVolumeClaim, in order to see if access to the snapshot is allowed.
If any ReferenceGrant allows access, the csi-provisioner provisions a volume from the snapshot.  
```

* Install the `Container Stroge Interface Driver` itself

## Putting it all together

To see how this works, you can install the sample and try it out.
This sample do to create PVC in ns1 namespace from VolumeSnapshot in default namespace by using csi-hostpath driver.

1. Deploy VolumeSnapshot controller.

   ```terminal
   kubectl apply -f https://raw.githubusercontent.com/kubernetes-csi/external-snapshotter/master/client/config/crd/snapshot.storage.k8s.io_volumesnapshotclasses.yaml
   kubectl apply -f https://raw.githubusercontent.com/kubernetes-csi/external-snapshotter/master/client/config/crd/snapshot.storage.k8s.io_volumesnapshotcontents.yaml
   kubectl apply -f https://raw.githubusercontent.com/kubernetes-csi/external-snapshotter/master/client/config/crd/snapshot.storage.k8s.io_volumesnapshots.yaml

   kubectl apply -f https://raw.githubusercontent.com/kubernetes-csi/external-snapshotter/master/deploy/kubernetes/snapshot-controller/rbac-snapshot-controller.yaml
   kubectl apply -f https://raw.githubusercontent.com/kubernetes-csi/external-snapshotter/master/deploy/kubernetes/snapshot-controller/setup-snapshot-controller.yaml

>    ```

Author: The description has been removed. Because it changed to prerequisite work.

2. Deploy External Provisioner controller.

   ```terminal
   kubectl apply -f https://raw.githubusercontent.com/kubernetes-csi/external-provisioner/master/deploy/kubernetes/deployment.yaml 
   kubectl apply -f https://raw.githubusercontent.com/kubernetes-csi/external-provisioner/master/deploy/kubernetes/rbac.yaml

>    ```

Reviewer 3: I recommend linking to a specific commit, and not to the _master_ branch. That keeps the hyperlink valid.

3. Deploy csi-hostpath driver.

   ```terminal
   cd /tmp
   git clone --depth=1 https://github.com/kubernetes-csi/csi-driver-host-path.git
   cd csi-driver-host-path
   ./deploy/kubernetes-latest/deploy.sh
   ```

4. Deploy ReferenceGrants CRD
  
   ```terminal

>    kubectl apply -f https://raw.githubusercontent.com/kubernetes-sigs/gateway-api/main/config/ crd/experimental/gateway.networking.k8s.io_referencegrants.yaml

Reviewer 3: I recommend linking to a specific commit, and not to the _main_ branch. That keeps the hyperlink valid.

   ```

5. Create StorageClass, PVC, and VolumeSnapshot by the examples in the csi-hostpath repo

>     ```terminal

Reviewer 3: (nit)
I'd highlight this as shell. Check that the indent is correct (should be 3 spaces).

    cd /tmp/csi-driver-host-path
    kubectl apply -f examples/csi-storageclass.yaml
    kubectl apply -f examples/csi-pvc.yaml

>     kubectl apply -f examples/csi-snapshot-v1beta1.yaml

Reviewer 4: just use examples/csi-snapshot-v1.yaml

>     ```

Reviewer 3: Can we avoid mentioning host path volumes? People use volume type even when it's not wise to.

Author: Is possible. In that case, it is assumed that there is a CSI driver setup, snapshot, etc. The implementation procedure does not include prerequisite work. Is this alright with you?

Reviewer 3: If you change `hostpath` to `example` then that's good for me. And then tell people they'll need to pick an actual storage driver if they want to follow along from home.

(aside: the text assumed that `/tmp/csi-driver-host-path` contains manifests, which it usually doesn't, so we were already assuming that people had extra steps to take).

Author: Thank you for advising.  I decided not to write the description of CSI-Hostpath.

6. Ceate a ns1 namespace

    ```terminal
    kubectl create ns ns1
    ```

> 7. Create a ReferenceGrant

Reviewer 3: ```suggestion
7. Create a ReferenceGrant. Here's a manifest for an example ReferenceGrant.
```

   ```yaml
   apiVersion: gateway.networking.k8s.io/v1alpha2
   kind: ReferenceGrant
   metadata:
     name: bar
     namespace: default
   spec:
     from:
     - group: ""
       kind: PersistentVolumeClaim
       namespace: ns1
     to:
     - group: snapshot.storage.k8s.io
       kind: VolumeSnapshot
       name: new-snapshot-demo
   ```

8. Create a PersistentVolumeClaim

   ```yaml
   apiVersion: v1
   kind: PersistentVolumeClaim
   metadata:
     name: foo-pvc
     namespace: ns1
   spec:

>      storageClassName: csi-hostpath-sc

Author: @sftim 
I have something I want to consult with you.
https://github.com/kubernetes/website/pull/37628/files#diff-3a022011d14568f9dcaaf14b8dce54ea2d72d3a496988e77ce7bb05df60cf6cdR32
This example assumes csi-hostpath. csi-storageclass.yaml also uses exmaples from csi-hostpath repo.
If you change to example, you will not be able to use the above file, so I think it will be a procedure for the user to create a storage class as a premise. Is this alright with you?

>      storageClassName: csi-hostpath-sc

Reviewer 3: If we don't change to a different example, we **must** warn readers about the risks of using `hostPath` in production environments.

     accessModes:
     - ReadWriteOnce
     resources:
       requests:
         storage: 1Gi
     dataSourceRef2:
       apiGroup: snapshot.storage.k8s.io
       kind: VolumeSnapshot
       name: new-snapshot-demo
       namespace: default
     volumeMode: Filesystem
   ```

Note This is only the simplest example.

## How can I learn more?

The enhancement proposal,
[Provision volumes from cross-namespace snapshots](https://github.com/kubernetes/enhancements/tree/master/keps/sig-storage/3294-provision-volumes-from-cross-namespace-snapshots), includes lots of detail about the history and technical implementation of this feature.

> Please get involved by joining the Kubernetes storage SIG to help us enhance this

Reviewer: ```suggestion
Please get involved by joining the Kubernetes SIG Storage to help us enhance this
```

Reviewer 3: Actually, “the Kubernetes storage SIG” is fine. Expanded out, it'd be “the Kubernetes storage special interest group” which is idiomatic.
